layers/shader_attribute_validation.h

```diff
@@ -0,0 +1,308 @@
+/* Copyright (c) 2015-2022 The Khronos Group Inc.
+ * Copyright (c) 2015-2022 Valve Corporation
+ * Copyright (c) 2015-2022 LunarG, Inc.
+ * Copyright (C) 2015-2022 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Author: Arno Galvez <arno@lunarg.com>
+ */
+
+#pragma once
+
+#include <cstdlib>
+#include <array>
+
+#include "shader_module.h"
+#include "range_vector.h"
+
+class ShaderAttributeValidation {
+  public:
+    enum class ComponentStatus {
+        Uninitialized,
+        Seen,
+        Unseen,
+    };
+
+    ShaderAttributeValidation(const SHADER_MODULE_STATE &shader_module,
+                              const std::map<location_t, interface_var>::const_iterator &shader_attributes_iterator,
+                              const std::map<location_t, interface_var>::const_iterator &shader_attributes_iterator_end)
+        : is_valid_(shader_attributes_iterator != shader_attributes_iterator_end) {
+        if (is_valid_) {
```

Comment:
Return early to avoid indenting the majority of the logic.

```suggestion
        if (!is_valid_) { return; }
```

Reply:
yes it looks better this way!